src/overlays/fbdemos/ovl_fbdemo_wipe3/z_fbdemo_wipe3.c

```diff
@@ -17,37 +19,141 @@ void TransitionWipe3_SetColor(void* thisx, u32 color);
+}
+
+void* TransitionWipe3_Init(void* thisx) {
+    TransitionWipe3* this = THIS;
+
+    bzero(this, sizeof(TransitionWipe3));
+    return this;
+}
+
+void TransitionWipe3_Destroy(void* thisx) {
+}
+
+void TransitionWipe3_UpdateScrollY(void* thisx, f32 scroll) {
+    TransitionWipe3* this = THIS;
+
+    this->scrollY = (s32)(500.0f * scroll);
+}
+
+void TransitionWipe3_Update(void* thisx, s32 updateRate) {
+    TransitionWipe3* this = THIS;
+
+    if (this->unk_14) {
+        if ((this->scrollY == 0) && (this->texIndex == 2)) {
+            func_8019F128(NA_SE_OC_SECRET_WARP_IN);
+        }
+        this->scrollY += (this->wipeSpeed * 3) / updateRate;
+        if (this->scrollY >= 500) {
+            this->scrollY = 500;
+            this->isDone = true;
+        }
+    } else {
+        this->scrollY -= (this->wipeSpeed * 3) / updateRate;
+        if (this->scrollY <= 0) {
+            this->scrollY = 0;
+            this->isDone = true;
+        }
+    }
+}
+
+void TransitionWipe3_Draw(void* thisx, Gfx** gfxP) {
+    Gfx* dl = *gfxP;
+    Mtx* modelView = &THIS->modelView[THIS->frame];
+    f32 scale = 14.8f;
+    Gfx* texScroll;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_fbdemo_wipe3/TransitionWipe3_Start.s")
+    THIS->frame ^= 1;
+    gDPPipeSync(dl++);
+    texScroll = Gfx_BranchTexScroll(&dl, THIS->scrollX, THIS->scrollY, 16, 64);
+    gSPSegment(dl++, 0x09, texScroll);
+    gSPSegment(dl++, 0x08, THIS->curTexture);
+    gDPSetColor(dl++, G_SETPRIMCOLOR, THIS->color.rgba);
+    gDPSetColor(dl++, G_SETENVCOLOR, THIS->color.rgba);
+    gSPMatrix(dl++, &THIS->projection, G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_PROJECTION);
+    gSPPerspNormalize(dl++, THIS->normal);
+    gSPMatrix(dl++, &THIS->lookAt, G_MTX_NOPUSH | G_MTX_MUL | G_MTX_PROJECTION);
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_fbdemo_wipe3/TransitionWipe3_Init.s")
+    if (scale != 1.0f) {
+        guScale(modelView, scale, scale, 1.0f);
+        gSPMatrix(dl++, modelView, G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
+    }
+    gSPDisplayList(dl++, sTransWipe3DL);
+    gDPPipeSync(dl++);
+    *gfxP = dl;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_fbdemo_wipe3/TransitionWipe3_Destroy.s")
+s32 TransitionWipe3_IsDone(void* thisx) {
+    TransitionWipe3* this = THIS;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_fbdemo_wipe3/func_80AC68B4.s")
+    return this->isDone;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_fbdemo_wipe3/TransitionWipe3_Update.s")
+void TransitionWipe3_SetType(void* thisx, s32 type) {
+    TransitionWipe3* this = THIS;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_fbdemo_wipe3/TransitionWipe3_Draw.s")
+    if (type & 0x80) {
+        this->unk_16 = (s8)(type & 1);
+        this->texIndex = (s8)((type >> 1) & 7);
+        this->unk_15 = (s8)((type >> 4) & 3);
+        return;
+    }
+    if (type == 1) {
+        this->unk_14 = true;
```

Comment:
This looks like the `direction`. Generally we create an enum instead of a boolean (see Wipe1).